jax_unirep/evotuning.py

```diff
@@ -314,37 +352,41 @@ def step(i, state):
     epoch_len = round(len(sequences) / batch_size)
 
     n = n_epochs * epoch_len
-    for i in range(n):
+    for i in tqdm(range(n), desc="Iteration"):
 
         if i % epoch_len == 0:
             logger.info(f"Starting epoch {int(i / epoch_len) + 1}")
-
-        logging.debug("Getting batches")
+        logger.debug(f"Iteration {i}")
+        logger.debug("Getting batches")
         l = choice(seq_lens)
         x, y = len_batching_funcs[l]()
 
         # actual forward & backwrd pass happens here
-        logging.debug("Getting state")
-        state = step(i, state)
+        logger.debug("Getting state")
+        state = step(i, x, y, state)
 
         if i % epoch_len == 0:
             params = get_params(state)
-            loss = avg_loss(xs, ys, params)
+            x, y = len_batching_funcs[l]()
+            loss = avg_loss([x], [y], params, backend=backend)
 
         if steps_per_print:
             if (i + 1) % (steps_per_print * epoch_len) == 0:
 
                 logger.info(
                     f"Epoch {int(i / epoch_len) + 1}: "
-                    f"Average training-set loss: {loss}. "
+                    f"Estimated average training-set loss: {loss}. "
                     f"Weights dumped."
                 )
 
-                if holdout_seqs is not None:
+                if holdout_seqs:
                     # calculate and print loss for out-domain holdout set
+                    l = choice(holdout_seq_lens)
+                    x, y = holdout_len_batching_funcs[l]()
+                    loss = avg_loss([x], [y], params, backend=backend)
```

Comment:
Same comment as above about what gets passed to `avg_loss`.

Reply:
Pushed something up. Let me know what you think about the new version.